src/sagas/walletConnect.js

```diff
@@ -0,0 +1,494 @@
+      topic: activeSessions[key].topic,
+      reason: {
+        code: ERROR_CODES.USER_DISCONNECTED,
+        message: '',
+      },
+    }));
+  }
+
+  yield call(refreshActiveSessions);
+}
+
+/**
+ * This saga will be called (dispatched from the event listener) when a session
+ * is requested from a dApp
+ */
+export function* onSessionRequest(action) {
+  const { payload } = action;
+  const { params } = payload;
+
+  const { web3wallet } = yield select((state) => state.walletConnect.client);
+  const activeSessions = yield call(() => web3wallet.getActiveSessions());
+  const requestSession = activeSessions[payload.topic];
+  if (!requestSession) {
+    console.error('Could not identify the request session, ignoring request..');
+    return;
+  }
+
+  const data = {
+    icon: get(requestSession.peer, 'metadata.icons[0]', null),
+    proposer: get(requestSession.peer, 'metadata.name', ''),
+    url: get(requestSession.peer, 'metadata.url', ''),
+    description: get(requestSession.peer, 'metadata.description', ''),
+  };
+
+  switch (params.request.method) {
+    case AVAILABLE_METHODS.HATHOR_SIGN_MESSAGE:
+      yield put({
+        type: 'SIGN_MESSAGE_REQUEST',
+        payload: {
+          ...data,
+          requestId: payload.id,
+          topic: payload.topic,
+          message: get(params, 'request.params.message'),
+        }
+      });
+      break;
+    default:
+      yield call(() => web3wallet.respondSessionRequest({
+        topic: payload.topic,
+        response: {
+          id: payload.id,
+          jsonrpc: '2.0',
+          error: {
+            code: ERROR_CODES.USER_REJECTED_METHOD,
+            message: 'Rejected by the user',
+          },
+        },
+      }));
+      break;
+  }
+}
+
+/**
+ * This saga will be called (dispatched from the event listener) when a sign
+ * message RPC is published from a dApp
+ */
+export function* onSignMessageRequest(action) {
+  const data = action.payload;
+  const { web3wallet } = yield select((state) => state.walletConnect.client);
+
+  const onAcceptAction = { type: 'WALLET_CONNECT_ACCEPT' };
+  const onRejectAction = { type: 'WALLET_CONNECT_REJECT' };
+
+  yield put(setWalletConnectModal({
+    show: true,
+    type: WalletConnectModalTypes.SIGN_MESSAGE,
+    data,
+    onAcceptAction,
+    onRejectAction,
+  }));
+
+  const { reject } = yield race({
+    accept: take(onAcceptAction.type),
+    reject: take(onRejectAction.type),
+  });
+
+  try {
+    if (reject) {
+      yield call(() => web3wallet.respondSessionRequest({
+        topic: data.topic,
+        response: {
+          id: data.id,
+          jsonrpc: '2.0',
+          error: {
+            code: ERROR_CODES.USER_REJECTED,
+            message: 'Rejected by the user',
+          },
+        },
+      }));
+      return;
+    }
+
+    const wallet = yield select((state) => state.wallet);
+
+    if (!wallet.isReady()) {
+      console.error('Got a session request but wallet is not ready, ignoring..');
```

Comment:
Shouldn't we do it before opening the modal? It's weird to show the modal, user accepts and you do nothing. 

The best would be to show a modal waiting for the wallet to get ready